File: internal/server/advanced_test.go
package server

import (
	"net/http"
	"strings"
	"testing"
	"time"

	"nib/internal/ceremony"
	"nib/internal/vault"
)

// The advanced-features switch (`/pending 451`). Dan, 2026-09-10: *"default off. menus hidden."*
//
// # The rule every arm below is about
//
// **Off has to mean the function stops.** A toggle that only hides UI is a lie the product tells,
// and this repo has shipped that exact shape twice — `/pending 378` found *"leave this ceremony"*
// releasing nothing, and `/pending 3` is open because discovery announces every 500 ms whether or
// not anyone is looking at the panel. So these drive the DOORS, not the panel.

// setAdvanced writes the four switches directly, which is the state a user would have reached
// through Settings.
func setAdvanced(t *testing.T, v *vault.Vault, a *vault.Advanced) {
	t.Helper()
	cur := v.Settings()
	cur.Advanced = a
	if err := v.SetSettings(cur); err != nil {
		t.Fatal(err)
	}
}

func TestTheDefaultIsOffAndOffRefusesAtTheRoute(t *testing.T) {
	ts, srv := startServerWith(t)
	c, csrf := authedClient(t, ts)
	v := srv.unlockedVault()

	// **Put the vault back to never-been-asked, because `authedClient` deliberately does not leave
	// it there.** That helper switches every advanced feature on so the package's thirty-eight
	// ceremony tests are about ceremonies rather than about this switch — and the price is that
	// THIS file has to restore the default explicitly. It is the right price: the default now has
	// exactly one place where it is written down and driven, instead of being an emergent property
	// of a helper nobody reads.
	setAdvancedNil(t, v)
	if v.Settings().Advanced != nil {
		t.Fatalf("setup: the vault still holds advanced settings %+v, so the refusals below would "+
			"not be about the default at all", v.Settings().Advanced)
	}

	// Timestamping: the route refuses before it reads a body.
	code, body := postForCode(t, c, csrf, ts.URL+"/api/timestamp", struct{}{})
	if code != http.StatusForbidden {
		t.Errorf("POST /api/timestamp with the feature unconfigured returned %d, want 403 — the "+
			"default is off, and off means the calendar servers are not contacted", code)
	}
	if body == "" {
		t.Error("the refusal carried no sentence; a user who forgot they turned it off learns nothing")
	}

	// The ceremony's two creating doors.
	if code, _ := postForCode(t, c, csrf, ts.URL+"/api/ceremony/convene", struct{}{}); code != http.StatusForbidden {
		t.Errorf("convene returned %d with ceremonies off, want 403", code)
	}
	if code, _ := postForCode(t, c, csrf, ts.URL+"/api/ceremony/accept", struct{}{}); code != http.StatusForbidden {
		t.Errorf("accept returned %d with ceremonies off, want 403", code)
	}
}

// TestSwitchingOnOpensTheSameDoors is the positive control, and without it every assertion above
// passes on a build that refuses everything unconditionally.
func TestSwitchingOnOpensTheSameDoors(t *testing.T) {
	ts, srv := startServerWith(t)
	c, csrf := authedClient(t, ts)
	setAdvanced(t, srv.unlockedVault(), &vault.Advanced{Ceremony: true, Timestamp: true})

	// Not 200 — these are empty bodies — but NOT 403 either: the feature gate is past, and what
	// answers now is the route's own validation.
	if code, _ := postForCode(t, c, csrf, ts.URL+"/api/timestamp", struct{}{}); code == http.StatusForbidden {
		t.Error("timestamping is switched ON and the route still refuses it as disabled")
	}
	if code, _ := postForCode(t, c, csrf, ts.URL+"/api/ceremony/convene", struct{}{}); code == http.StatusForbidden {
		t.Error("ceremonies are switched ON and convene still refuses as disabled")
	}
}

// TestTheArmSweepStopsWhenCeremoniesAreOff — the door a route-only gate would miss.
//
// D14 made accepting an invitation arm, and `rearmCeremonies` renews that arm at every unlock. A
// switch that refused only convene and accept would leave the machine listening for hops on a
// feature its user had turned off, which is the hidden-button-only shape verbatim.
func TestTheArmSweepStopsWhenCeremoniesAreOff(t *testing.T) {
	ts, srv := startServerWith(t)
	srv.EnableDeliveryRearm()
	c, csrf := authedClient(t, ts)
	v := srv.unlockedVault()
	setAdvanced(t, v, &vault.Advanced{Ceremony: true, Discovery: true, Rendezvous: true})

	me := myFingerprint(t, c, ts.URL)
	invitation, _ := inviteFor(t, me)
	if code, body := postForCode(t, c, csrf, ts.URL+"/api/ceremony/accept",
		acceptRequest{Invitation: invitation}); code != http.StatusOK {
		t.Fatalf("setup: accept returned %d: %s", code, body)
	}
	// SETUP: the arm actually comes up with the feature ON. Without this, "not armed when off" is
	// satisfied by a machine that never arms at all — the vacuous green in its plainest form.
	if !armedWithin(srv, 3*time.Second) {
		t.Fatal("setup: accepting did not arm with ceremonies switched ON, so this test cannot " +
			"show that switching them off is what stops it")
	}
	if code, body := postForCode(t, c, csrf, ts.URL+"/api/session/disarm", struct{}{}); code != http.StatusOK {
		t.Fatalf("setup: disarm returned %d: %s", code, body)
	}

	setAdvanced(t, v, &vault.Advanced{})
	srv.rearmCeremonies(v)
	if srv.sess.status().Armed {
		postForCode(t, c, csrf, ts.URL+"/api/session/disarm", struct{}{})
		t.Error("the sweep armed for a ceremony on a machine whose user has switched ceremonies " +
			"off — the feature is still running and only its panel is gone")
	}
}

// TestSwitchingCeremoniesOffIsRefusedWhileOneIsLive.
//
// `Termination` carries two attested end states and `Receipt` two derived ones, and *"the user
// switched the feature off"* is none of them: ending a live proceeding this way would have to be
// written as `abandoned` or `stopped`, both false statements to every other party. Refusing is the
// only outcome the record can express.
func TestSwitchingCeremoniesOffIsRefusedWhileOneIsLive(t *testing.T) {
	ts, srv := startServerWith(t)
	c, csrf := authedClient(t, ts)
	v := srv.unlockedVault()
	setAdvanced(t, v, &vault.Advanced{Ceremony: true})
	rec, _, _ := ceremonyOnDisk(t)

	// SETUP: the ceremony really is live, or the refusal below would be about nothing.
	if st := ceremony.ReadStored(defaultOutputDir(), rec.ID, time.Now()); st.Ended != "" {
		t.Fatalf("setup: the fixture ceremony is already ended (%q)", st.Ended)
	}

	off := advancedRequest{Ceremony: false}
	code, body := postForCode(t, c, csrf, ts.URL+"/api/settings", settingsRequest{Advanced: &off})
	if code != http.StatusConflict {
		t.Errorf("switching ceremonies off with one still running returned %d, want 409 — the "+
			"proceeding would be left unreachable with no state the record can express for it", code)
	}
	if !strings.Contains(body, "has not finished") {
		t.Errorf("the refusal says %q and does not tell the user what is running", body)
	}
	// And nothing was written: a refused change that stored itself anyway is the worst of both.
	if a := v.Settings().Advanced; a == nil || !a.Ceremony {
		t.Errorf("the refused change was stored anyway (%+v)", a)
	}
}

// TestTheSeedLeavesALiveCeremonyReachable — the migration Dan's default-off choice creates.
func TestTheSeedLeavesALiveCeremonyReachable(t *testing.T) {
	// A vault that has never been asked, on a machine that already holds a live ceremony: exactly
	// what an existing install looks like the first time it runs this build.
	sv := seedableVault(t)
	ceremonyOnDisk(t)
	setAdvancedNil(t, sv)

	seedSrv.SeedAdvanced(sv)
	a := sv.Settings().Advanced
	if a == nil {
		t.Fatal("the seed wrote nothing, so it will run again at every unlock and a user who " +
			"switches the ceremony off will find it back on after a restart")
	}
	if !a.Ceremony || !a.Discovery || !a.Rendezvous {
		t.Errorf("the seed left a live ceremony stranded: %+v. Its hop can never arrive, the "+
			"panel that would say so is hidden, and the user has no way to find out why", a)
	}
	if a.Timestamp {
		t.Error("the seed switched timestamping on. There is no on-disk state saying anyone was " +
			"using it, so nothing is stranded by leaving it off — turning it on is a second " +
			"default rather than a migration")
	}
}

// TestTheSeedLeavesAQuietMachineOff is the other half, and it is the common case: an existing
// vault with no live proceeding gets the default Dan asked for, not an exemption from it.
func TestTheSeedLeavesAQuietMachineOff(t *testing.T) {
	sv := seedableVault(t)
	setAdvancedNil(t, sv)
	seedSrv.SeedAdvanced(sv)
	a := sv.Settings().Advanced
	if a == nil {
		t.Fatal("the seed wrote nothing")
	}
	if a.Ceremony || a.Discovery || a.Rendezvous || a.Timestamp {
		t.Errorf("a machine with no live ceremony was seeded %+v — 'default off' became 'off for "+
			"new installs only', which is not what was asked for", a)
	}
}

// TestTheSeedDoesNotUndoTheUsersChoice.
func TestTheSeedDoesNotUndoTheUsersChoice(t *testing.T) {
	sv := seedableVault(t)
	ceremonyOnDisk(t) // live, so the seed WOULD turn the ceremony on if it ran
	setAdvanced(t, sv, &vault.Advanced{})

	seedSrv.SeedAdvanced(sv)
	if a := sv.Settings().Advanced; a == nil || a.Ceremony {
		t.Errorf("the seed overwrote a user who had switched everything off (%+v) — a setting "+
			"that undoes itself at every restart is worse than no setting", a)
	}
}

// seedSrv is the server whose SeedAdvanced is under test. `SeedAdvanced` reads only the vault it is
// handed and `hasLiveCeremony`'s HOME, so any enrolled server serves.
var seedSrv *Server

// seedableVault enrols a real vault and hands it back, with HOME already a sandbox.
//
// **It replaced a `t.Skip` on `srv.unlockedVault() == nil`, which was silently true for all three
// seed tests**: `startServerWith` builds a server and `authedClient` is what enrols one. Three tests
// reported SKIP and nothing was checked — the same shape as a guard whose scan finds no files.
func seedableVault(t *testing.T) *vault.Vault {
	t.Helper()
	ts, srv := startServerWith(t)
	authedClient(t, ts)
	v := srv.unlockedVault()
	if v == nil {
		t.Fatal("enrolling did not leave the server with an unlocked vault")
	}
	seedSrv = srv
	return v
}

// setAdvancedNil puts the vault back to never-been-asked.
func setAdvancedNil(t *testing.T, v *vault.Vault) {
	t.Helper()
	cur := v.Settings()
	cur.Advanced = nil
	if err := v.SetSettings(cur); err != nil {
		t.Fatal(err)
	}
}

File: test/jsdom/advanced.test.mjs
// Advanced features, at the surface (`/pending 451`). Dan, 2026-09-10: *"default off. menus hidden."*
//
// ── The half this tier owns, and the half it must not claim ──────────────────
// **This file proves the HIDING and nothing else.** The Go tests prove the switch stops the
// function — the announcer's socket, the DHT bootstrap, the arm sweep, the two timestamp routes —
// and that half is the one that matters: a toggle which only hid UI would be a lie the product
// tells, and this repo has shipped that shape twice (`/pending 378`, `/pending 3`). So nothing here
// asserts that anything stopped; it asserts that what is offered matches what the machine will do.
//
// ── The collision this file is the guard for ─────────────────────────────────
// "Off ⇒ hidden" cuts at PANEL granularity, never at the mode. The ceremony lives inside the
// `collaborate` mode, which also carries co-signing and Simple Sign — hiding the mode tab would
// take ordinary two-party signing with it. That is the failure an item-by-item design round is
// structurally blind to, and it is asserted below.
//
// One boot per file — see boot.mjs.
import test from 'node:test';
import assert from 'node:assert/strict';
import { boot } from './boot.mjs';

let saved = null;
let refuse = null;

const h = await boot({
  routes: {
    // **Its own status, with the features OFF.** boot.mjs's default has them on so that every
    // other file here is about its own subject rather than about this switch; this file is about
    // the switch, so it states the default explicitly. One place where "off is the default" is
    // written down and driven, rather than an emergent property of a shared fixture.
    '/api/status': {
      state: 'ready', csrf: 'test-csrf', version: 'test',
      autoUpdate: false, updateCheckLocked: false, ghostscript: false, libreoffice: false,
    },
    '/api/settings': (opts) => {
      saved = JSON.parse((opts && opts.body) || '{}');
      if (refuse) return new Response(JSON.stringify({ error: refuse }), { status: 409, headers: { 'Content-Type': 'application/json' } });
      return { status: 'ok' };
    },
  },
});
const { document: doc, settle } = h;

const shown = (sel) => { const el = doc.querySelector(sel); return !!el && !el.hidden; };
const box = (id) => doc.getElementById(id);

test('the default is off, and the card says so with four unticked boxes', () => {
  for (const id of ['advCeremonyChk', 'advDiscoveryChk', 'advRendezvousChk', 'advTimestampChk']) {
    assert.ok(box(id), `${id} is missing — the switch has no surface at all`);
    assert.equal(box(id).checked, false,
      `${id} is ticked on a machine whose vault holds no advanced settings. Off is the default, `
      + 'and a box that says otherwise is the one place a user checks before trusting it');
  }
});

test('a feature that is off is not offered', () => {
  assert.equal(shown('#ceremony'), false,
    'the Signing Ceremonies panel is offered while ceremonies are switched off — the product is '
    + 'inviting a user into a flow the server will refuse with 403');
  assert.equal(shown('#timestampBtn'), false, 'Timestamp is offered while timestamping is off');
  assert.equal(shown('#timestampVerifyBtn'), false, 'Verify a timestamp is offered while it is off');
});

test('hiding the ceremony does NOT take ordinary co-signing with it', () => {
  // The collision. `collaborate` is a whole mode and the ceremony is one panel inside it; cutting
  // at the mode would remove Simple Sign and Send & Receive, which are not the ceremony.
  // **`shown`, not `querySelector`.** The first cut asserted these elements EXIST, and
  // `querySelector` finds hidden ones — so a mutation that cut at mode granularity, hiding the
  // Signing tab and both co-sign buttons, left this test entirely green. Existence was never the
  // claim; being offered is.
  assert.equal(shown('.modetab[data-tab="collaborate"]'), true,
    'the Signing MODE TAB was hidden along with the ceremony panel. Co-signing, Simple Sign and '
    + 'Send & Receive live in that mode and are not the feature that was switched off');
  assert.equal(shown('#cosignBtn'), true, 'Co-sign with a peer went with the ceremony');
  assert.equal(shown('#sessionInitBtn'), true, 'Co-sign live with a peer went with the ceremony');
});

test('switching one on sends all four, and reveals its surface', async () => {
  saved = null;
  box('advCeremonyChk').checked = true;
  box('advCeremonyChk').onchange();
  await settle();
  assert.ok(saved && saved.advanced, 'ticking a box sent no advanced object');
  assert.deepEqual(saved.advanced, { ceremony: true, discovery: false, rendezvous: false, timestamp: false },
    'the request did not carry all four. A partial update cannot express "configured, all off" — '
    + 'four absent fields is what a client sends when it is changing something else entirely');
  assert.equal(shown('#ceremony'), true,
    'the ceremony panel is still hidden after switching ceremonies on, so the only way to reach '
    + 'the feature is a restart');
});

test('a refusal puts the box back, rather than showing a state the machine is not in', async () => {
  // The server refuses switching ceremonies off while a proceeding is live (409). Echoing the
  // REQUEST would leave the box saying off while the feature is on — the control and the thing it
  // controls disagreeing, which is worse than either state.
  refuse = 'a signing ceremony on this machine has not finished';
  box('advCeremonyChk').checked = false;
  box('advCeremonyChk').onchange();
  await settle();
  assert.equal(box('advCeremonyChk').checked, true,
    'the box stayed unticked after the server refused the change — it now reports a setting this '
    + 'machine does not hold, and the feature is still running');
  assert.equal(shown('#ceremony'), true, 'the panel was hidden on a change the server refused');
  const err = doc.getElementById('advError');
  assert.ok(err && !err.hidden, 'the refusal was silent; the user sees a box spring back and no reason');
  assert.match(err.textContent, /not finished/,
    `the error reads ${JSON.stringify(err.textContent)} and does not say what is running`);
  refuse = null;
});
